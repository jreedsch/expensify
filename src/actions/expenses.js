// ACTION GENERATORS FOR STORE EXPENSES ATTRIBUTE
import uuid from 'uuid';
import database from '../firebase/firebase';

// return function, not object
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database.ref('/expenses').push(expense) //return is for promise chaining in test
    .then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  }
};

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const removeExpense = ({ id } = {}) => {
  return {
    type: 'REMOVE_EXPENSE',
    id
  }
}

export const editExpense = (id, updates) => {
  console.log("IN action creator, id: "+id+", updates: "+JSON.stringify(updates));
  return (
    { type: 'EDIT_EXPENSE',
      id,
      updates
    }
  );
};

//export const editExpense = (id, updates) => ({
//  type: 'EDIT_EXPENSE',
//  id,
//  updates
//});
