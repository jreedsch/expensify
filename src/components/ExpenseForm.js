import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css';

//const now = moment();  today
//console.log(now.format('MMM Do, YYYY'));

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100) : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      err: ''
    };

  };

  onDescriptionChange = (e) => {
    const description = e.target.value; // or use event.persist()
    this.setState(() => ({ description }));  //es6 object shorthand
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));  //es6 object shorthand
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    const currencyRegex = '/^\d{1,}(\.\d{0,2})$/'; // ^ begins with
                                              // \d decimal
                                              // * any count, {1,} 1 to any
                                              // () optional
                                              // decimal point (escaped)
                                              // {0,2} from 0 to 2 decimals
                                              // $ end
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));  //es6 object shorthand
    } else {
      console.log("amount does not match regex filter: "+amount);
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onCalendarFocusChange = ({focused}) => {
    this.setState(() => ({ calendarFocused: focused}));
  };



  onSubmitHandler = (e) => {
    console.log("SUBMIT, description="+this.state.description);
    e.preventDefault(); //no refresh
    if (!this.state.description || !this.state.description.length > 0) {
    //if (!this.state.description || !this.state.amount > 0) {
      this.setState(() => ({ err: 'description is required' }));
    } else if (!this.state.amount || !this.state.amount > 0) {
    //if (!this.state.description || !this.state.amount > 0) {
      this.setState(() => ({ err: 'amount is required' }));
    } else {
      this.setState(() => ({ err: '' }));
      console.log("IN ExpenseForm, state: "+JSON.stringify(this.state));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }

  }

  render() {
    return (

        <form className="form" onSubmit={this.onSubmitHandler}>
          {this.state.err && <p className="form__error">{this.state.err}</p> }
          <input
            type='text'
            className="text-input"
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type='text'
            placeholder='Amount'
            className="text-input"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            className="textarea-input"
            placeholder='Note'
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <div> { /*  removes from form children (_form.scss: > * { lecture #176 ), so not full width  */ }
            <span><button type="submit" >Save Expense</button></span>
            <span><button type="button" className='cancel_button' onClick={this.props.onCancel}>Cancel</button></span>
          </div>
        </form>
    )

  }
}

export default ExpenseForm ;
