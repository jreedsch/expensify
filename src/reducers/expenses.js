const expensesReducerDefaultState = [];

//export default (state = expensesReducerDefaultState, action) => {
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  console.log("IN expense reducer, state: "+JSON.stringify(state));
  console.log("IN expense reducer, action: "+JSON.stringify(action));
  switch (action.type) {
    case 'ADD_EXPENSE':
      console.log("IN expenses reducer, add expense: ");
      console.log({...action.expense});
      return [ ...state, action.expense ] //state.concat(action.expense);

    case 'SET_EXPENSES': //wipe state expenses
      console.log("IN expenses reducer SET_EXPENSES, expenses: ");
      console.log({...action.expenses});
      return action.expenses;  //[ ...state, ...action.expenses ]
      //return state;

    case 'LOAD_EXPENSES': //wipe state expenses
      console.log("IN expenses reducer LOAD_EXPENSES, expenses: ");
      console.log({...action.expenses});
      return action.expenses;  //[ ...state, ...action.expenses ]

    case 'REMOVE_EXPENSE':
      console.log("IN expenses reducer REMOVE_EXPENSE, id: "+action.id);
      //return state.filter((expense) => { return expense.id !== action.expense.id});
      //return state.filter(( {id} ) => { return id !== action.id});
      return state.filter(( {id} ) => id !== action.id );

    case 'EDIT_EXPENSE':
      return state.map((expense) => {
          if (expense.id === action.id) {
            console.log("IN expense reducer, expense: "+JSON.stringify(expense));
            return {...expense, ...action.updates}; //override existing attrs with updates object attrs
          } else {
            return expense;
          }
        }
      );
    default:
      return state;
  }
};

export default expensesReducer;
