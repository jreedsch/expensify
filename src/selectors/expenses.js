import moment from 'moment';

// filter views
const getVisibleExpensesX = (expenses, filters) => {
  console.log("IN getVisibleExpenses, expenses: "+expenses);
  console.log("IN getVisibleExpenses, filters: "+filters);
}

export const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
      //console.log("expense: "+JSON.stringify(expense));
      //console.log("startDate: "+startDate);
      //console.log("endDate: "+endDate);
      //console.log("expense.createdAt: "+expense.createdAt);
      //console.log("text: "+text);
      //console.log("expense.description.toLowerCase(): "+expense.description.toLowerCase());
      //const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate; //disregard if not a number
      //const endDateMatch   = typeof endDate !== 'number' || expense.createdAt <= endDate;

      const createdAtMoment = moment(expense.createdAt);

      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch   = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;


      let textMatch        = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

      //console.log("textMatch: "+textMatch);

      return startDateMatch && endDateMatch && textMatch;
    }
  ).sort((a,b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount'){
      return a.amount < b.amount? 1 : -1;
    }
  })
}

//export default getVisibleExpenses;

// sum up all expense amounts
export const getExpenseTotal = (expenses) => {

  //const expenseAmount = expenses.map((expense) => {
  //  return expense.amount;
  //})
  //return expenseAmount.reduce((acc, val) => {
  //  return acc + val;
  //}, 0);

  return expenses
         .map((expense) => expense.amount)
         .reduce((sum,value) => sum + value, 0);
}
