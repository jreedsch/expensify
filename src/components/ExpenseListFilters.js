import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount,sortByDate, setStartDate, setEndDate } from  '../actions/filters';
import { DateRangePicker } from 'react-dates';

// dispatch is set on props by react-redux
export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    console.log("IN ExpenseListFilters.onFocusChange");
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    //const ndx = e.target.selectedIndex; //can use e.target.value
    //if (ndx === 0) {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">

          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="Search descriptions"
              type='text'
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>

          <div className="input-group__item">
            <select
              className="text-input"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value='date'>Date</option>
              <option value='amount'>Amount</option>
            </select>
          </div>

          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    ); // end return
  }; //end render
}; //end class

const mapDispatchToProps = (dispatch, props) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

const mapStateToProps = (state) => ( {filters: state.filters} ); //implicit return
//const mapStateToProps = (state) => {   //explicit return
//  return {
//    filters: state.filters
//  };
//};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
