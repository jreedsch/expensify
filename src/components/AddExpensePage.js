import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses'; //addExpense

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    console.log("IN AddExpensePage, add submitted.");
    this.props.startAddExpense(expense); //props.dispatch(addExpense(expense)); see mapDispatchToProps
    this.props.history.push('/');
  };

  onCancelButtonClick = (event) => {
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Add Expense</h1>
          </div>
        </div>

        <div className="content-container">
          <ExpenseForm
            onSubmit={this.onSubmit}
            onCancel={this.onCancelButtonClick}
          />
        </div>

      </div>
    );
  }
}

/*
const AddExpensePage = (props) => ( //implicit return
  <div>
    <h3>add expense page</h3>
    <ExpenseForm onSubmit={(expense) => {
      console.log("IN AddExpensePage, add submitted.");
      props.onSubmit(expense); //props.dispatch(addExpense(expense)); see mapDispatchToProps
      props.history.push('/');
    }}/>
  </div>
);
*/

// makes testing the dispatch much easier
const mapDispatchToProps = (dispatch) => ({startAddExpense: (expense) => dispatch(startAddExpense(expense)) })
//{
//  return {
//    onSubmit: (expense) => dispatch(addExpense(expense))
//  }
//};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
