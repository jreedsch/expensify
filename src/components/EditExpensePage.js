import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {

  onSubmitEdit = (expense) => {
    console.log("IN EditExpensePage, edit submitted, expense: "+JSON.stringify(expense));
    console.log("this.props 1: "+JSON.stringify(this.props));
    console.log("edit expense to submit:  "+JSON.stringify(expense));
    console.log("edit expense id to submit:  "+this.props.expense.id);
    this.props.editExpense(this.props.expense.id, expense); //(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemoveButtonClick = (event) => {
    console.log("this.props 2: "+JSON.stringify(this.props));
    this.props.removeExpense({id: this.props.expense.id}); //this.props.removeExpense(this.props.expense.id)
    this.props.history.push('/');
  };

//        <p> edit expense page for id {this.props.match.params.id}</p>
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmitEdit}
        />
        <button onClick={ this.onRemoveButtonClick } >Remove!</button>
      </div>
    )
  };
}

const mapDispatchToProps = (dispatch, props) => ({
  removeExpense: (data) => dispatch(startRemoveExpense(data)), //dispatch(removeExpense(data)),
  editExpense: (id, expense) => dispatch(startEditExpense(id, expense))
});
  //removeExpense: (id) => dispatch(removeExpense({ id })) }

//const mapDispatchToProps = (dispatch) => ([
//  {editExpense: (id, expense) => dispatch(editExpense(id, expense)) },
//  {removeExpense: (id) => dispatch(removeExpense({ id })) }
//]);

/*
const EditExpensePage = (props) => {
  console.log("IN EditExpensePage, props: "+JSON.stringify(props)); //props from react-router
  return (
    <div>
      <p> edit expense page for id {props.match.params.id}</p>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          console.log("IN EditExpensePage, edit submitted, expense: "+JSON.stringify(expense));
          props.dispatch(editExpense(props.expense.id, expense));
          props.history.push('/');
        }}
      />
      <button onClick={(event) => {
        props.dispatch(removeExpense({id: props.expense.id}));
        props.history.push('/');
      }}>Remove!</button>
    </div>
  );
};

*/

//$$$ const mapStateToProps = (state, props) => ( { expense: state.expenses.find((expense) =>  expense.id === props.match.params.id) } );
const mapStateToProps = (state, props) => {
  const editExpense = state.expenses.find((expense) =>  expense.id === props.match.params.id);
  console.log("IN EditExpensePage.mapStateToProps, state: "+JSON.stringify(state));
  console.log("IN EditExpensePage.mapStateToProps, props: "+JSON.stringify(props));
  console.log("IN EditExpensePage.mapStateToProps, editExpense: "+JSON.stringify(editExpense));
  return { expense:  editExpense };
};



/*
const mapStateToProps = (state, props) => {
  return {
    //expense: state.expenses.find((expense) => {
    //  return expense.id === props.match.params.id;
    //})
    expense: state.expenses.find((expense) =>  expense.id === props.match.params.id)
  };
};
*/

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage) ;
