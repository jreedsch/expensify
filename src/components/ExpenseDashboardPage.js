import React from 'react';
import ExpenseListPage from './ExpenseListPage';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDashboardPage = () => ( //implicit return
  <div>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseListPage />
    <p style={styles} >src/components/ExpenseDashboardPage.js</p>
  </div>
);

const styles = {
  fontSize: '10px',
  float: 'right'
}
export default ExpenseDashboardPage ;
