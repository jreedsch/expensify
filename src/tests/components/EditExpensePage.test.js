import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage'; //use named export for unconnected component
import { expensesCase1 } from '../fixtures/expenses';

let editExpense, removeExpense, history, wrapper, expenseData;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  expenseData = expensesCase1[1];
  wrapper = shallow(<EditExpensePage
              editExpense={editExpense}
              removeExpense={removeExpense}
              history={history}
              expense={expenseData}/>);
});

test('should render edit expense page correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle submit of edit expense data', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenseData);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenseData.id, expenseData);
});

test('should handle remove of an expense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({id: expenseData.id});
});

/*
test('should handle submit of add expense data', () => {
  //const onSubmit = jest.fn();
  //const history = { push: jest.fn() };
  const expenseData = expenses[1];
  //const wrapper = shallow(<AddExpensePage onSubmit={onSubmit} history={history} />);
  wrapper.find('ExpenseForm').prop('onSubmit')(expenseData);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(addExpense).toHaveBeenLastCalledWith(expenseData);
});

*/
/*
test('should ', () => {
  const wrapper = shallow();
  expect(wrapper).
})
*/
