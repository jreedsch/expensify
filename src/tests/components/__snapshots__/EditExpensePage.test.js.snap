// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 200,
          "createdAt": -345600000,
          "description": "milk",
          "id": "2",
          "note": "",
        }
      }
      onCancel={[Function]}
      onSubmit={[Function]}
    />
    <button
      className="remove_button"
      onClick={[Function]}
    >
      Remove this expense!
    </button>
  </div>
</div>
`;
